src/thread.rs

```diff
@@ -50,23 +54,106 @@ pub fn spawn<F>(f: F) -> ThreadHandle
 where
     F: Send + 'static + FnOnce(mio::Poll) -> (),
 {
-    let poller = mio::Poll::new().unwrap();
-    let (registration, readiness) = mio::Registration::new2();
-
+    let child_poller = mio::Poll::new().unwrap();
+    let (shutdown_event_registration, shutdown_event) = mio::Registration::new2();
     ThreadHandle {
-        readiness: readiness,
-
+        shutdown_event: shutdown_event,
         handle: thread::spawn(move || {
-            poller
+            child_poller
                 .register(
-                    &registration,
+                    &shutdown_event_registration,
                     constants::SYSTEM,
                     mio::Ready::readable(),
                     mio::PollOpt::edge(),
                 )
                 .expect("Failed to register system pipe");
 
-            f(poller);
+            f(child_poller);
         }),
     }
 }
+
+/// mio Eventable ThreadPool.
+pub struct ThreadPool {
+    /// thread_id counter.
+    thread_id: usize,
+    /// Listing of all the joinable threads in the pool.
+    joinable: sync::Arc<sync::Mutex<Vec<usize>>>,
+    /// Mapping of thread_id to ThreadHandle.
+    threads: util::HashMap<usize, ThreadHandle>,
+    /// Mio readiness flagging when threads finish execution.
+    thread_event_readiness: option::Option<mio::SetReadiness>,
+}
+
+
+impl ThreadPool {
+
+    /// Construct a new ThreadPool.
+    pub fn new(thread_events_readiness: option::Option<mio::SetReadiness>) -> Self {
+        ThreadPool {
+            thread_id: 0,
+            joinable: sync::Arc::new(sync::Mutex::new(Vec::new())),
+            thread_event_readiness: thread_events_readiness,
+            threads: util::HashMap::default(),
+        }
+    }
+
+    /// Spawn a new thread and assign it to the pool.
+    pub fn spawn<F>(&mut self, f: F) -> usize
+    where
+        F: Send + 'static + FnOnce(mio::Poll) -> (),
+    {
+        let id = self.next_thread_id();
+        let joinable_arc = self.joinable.clone();
+        let thread_event_readiness = self.thread_event_readiness.clone();
+        let handler =
+            spawn(move |poller| {
+                f(poller);
+
+                let mut joinable = joinable_arc.lock().unwrap();
+                joinable.push(id);
+
+                if let Some(readiness) = thread_event_readiness {
+                    readiness
+                        .set_readiness(mio::Ready::readable())
+                        .expect("Failed to flag readiness for ThreadPool event!");
+                }
+            });
+        self.threads.insert(id, handler);
+        id
+    }
+
+    fn next_thread_id(&mut self) -> usize
+    {
+        let thread_id = self.thread_id;
+        self.thread_id += 1;
+        thread_id
+    }
+
+    /// Block on completion of all executing threads.
+    pub fn join(mut self) -> Vec<usize>
+    {
+        self.threads.drain().for_each(|(_, h)| h.join());
+        self.join_ready()
+    }
```

Comment:
This function looks incorrect... it looks like shutdown does the correct thing. `self.threads.drain` will empty the container, but then `join_ready` will attempt to `let handle = self.threads.remove(&id).unwrap(); handle.join()` below.

Reply:
Good catch.  I will wrap the remove in a guard.

Reply:
Updated.